File: src/typescript/src/transport/in-memory/in-memory-transport.ts
import { IncomingMessage } from '../transaction/message/incoming-message';
import { IncomingTransaction } from '../transaction/incoming-transaction';
import { IPolyBus } from '../../i-poly-bus';
import { ITransport } from '../i-transport';
import { MessageInfo } from '../transaction/message/message-info';
import { OutgoingMessage } from '../transaction/message/outgoing-message';
import { PolyBusBuilder } from '../../poly-bus-builder';
import { Transaction } from '../transaction/transaction';

export class InMemoryTransport {
  public addEndpoint(_builder: PolyBusBuilder, bus: IPolyBus): ITransport {
    const endpoint = new Endpoint(this, bus);
    this._endpoints.set(bus.name, endpoint);
    return endpoint;
  }
  private readonly _endpoints = new Map<string, Endpoint>();

  public async send(transaction: Transaction): Promise<void> {
    if (!this._active) {
      throw new Error('Transport is not active.');
    }

    if (transaction.outgoingMessages.length === 0) {
      return;
    }

    let promiseResolver: () => void = () => {};
    const transactionId = Math.random().toString(36).substring(2, 11);
    const promise = new Promise<void>((resolve) => {
      promiseResolver = resolve;
    });
    this._transactions.set(transactionId, promise);

    try {
      const tasks: Promise<void>[] = [];

      for (const message of transaction.outgoingMessages) {
        if (message.deliverAt) {
          let delay = message.deliverAt.getTime() - Date.now();
          if (delay > 0) {
            // eslint-disable-next-line no-undef
            const timeoutId = setTimeout(async () => {
              try {
                const transaction = await message.bus.createTransaction();
                message.deliverAt = undefined;
                transaction.outgoingMessages.push(message);
                await transaction.commit();
              } catch (error) {
                // eslint-disable-next-line no-undef
                console.error('Error delivering delayed message:', error);
              } finally {
                this._timeouts.delete(transactionId);
              }
            }, delay);
            this._timeouts.set(transactionId, timeoutId);
            continue;
          }
        }
        for (const endpoint of this._endpoints.values()) {
          const task = endpoint.handle(message);
          tasks.push(task);
        }
      }

      await Promise.all(tasks);
    } finally {
      promiseResolver();
      this._transactions.delete(transactionId);
    }
  }
  // eslint-disable-next-line no-undef
  private readonly _timeouts = new Map<string, NodeJS.Timeout>();

  public useSubscriptions: boolean = false;

  public async start(): Promise<void> {
    this._active = true;
  }
  private _active: boolean = false;

  public async stop(): Promise<void> {
    this._active = false;
    for (const timeoutId of this._timeouts.values()) {
      // eslint-disable-next-line no-undef
      clearTimeout(timeoutId);
    }
    this._timeouts.clear();
    await Promise.all(this._transactions.values());
    this._transactions.clear();
  }
  private readonly _transactions = new Map<string, Promise<void>>();
}

class Endpoint implements ITransport {
  constructor(
    private readonly transport: InMemoryTransport,
    private readonly bus: IPolyBus
  ) {}

  public async handle(message: OutgoingMessage): Promise<void> {
    if (!this.transport.useSubscriptions || this._subscriptions.includes(message.messageType)) {
      const incomingMessage = new IncomingMessage(this.bus, message.body);
      incomingMessage.headers = message.headers;

      try {
        const transaction = await this.bus.createTransaction(incomingMessage) as IncomingTransaction;
        await transaction.commit();
      } catch (error) {
        console.error('Error processing message:', error);
      }
    }
  }

  private readonly _subscriptions: Function[] = [];
  public async subscribe(messageInfo: MessageInfo): Promise<void> {
    const type = this.bus.messages.getTypeByMessageInfo(messageInfo);
    if (!type) {
      throw new Error(`Message type for attribute ${messageInfo.toString()} is not registered.`);
    }
    this._subscriptions.push(type);
  }

  public get supportsCommandMessages(): boolean {
    return true;
  }

  public get supportsDelayedMessages(): boolean {
    return true;
  }

  public get supportsSubscriptions(): boolean {
    return true;
  }

  public async send(transaction: Transaction): Promise<void> {
    return this.transport.send(transaction);
  }

  public async start(): Promise<void> {
    return this.transport.start();
  }

  public async stop(): Promise<void> {
    return this.transport.stop();
  }
}

File: src/dotnet/PolyBus/Transport/InMemory/InMemoryTransport.cs
using Microsoft.Extensions.Logging;
using PolyBus.Transport.Transactions;
using PolyBus.Transport.Transactions.Messages;

namespace PolyBus.Transport.InMemory;

public class InMemoryTransport
{
    public ITransport AddEndpoint(PolyBusBuilder builder, IPolyBus bus)
    {
        var endpoint = new Endpoint(this, bus);
        _endpoints.Add(bus.Name, endpoint);
        return endpoint;
    }
    readonly Dictionary<string, Endpoint> _endpoints = [];
    class Endpoint(InMemoryTransport transport, IPolyBus bus) : ITransport
    {
        public async Task Handle(OutgoingMessage message)
        {
            if (!transport.UseSubscriptions || _subscriptions.Contains(message.MessageType))
            {
                var incomingMessage = new IncomingMessage(bus, message.Body)
                {
                    Headers = message.Headers
                };

                try
                {
                    var transaction = (IncomingTransaction)await bus.CreateTransaction(incomingMessage);

                    await transaction.Commit();
                }
                catch (Exception error)
                {
                    var logger = incomingMessage.State.Values.OfType<ILogger>().FirstOrDefault();
                    logger?.LogError(error, error.Message);
                }
            }
        }

        readonly List<Type> _subscriptions = [];
        public Task Subscribe(MessageInfo messageInfo)
        {
            var type = bus.Messages.GetTypeByMessageInfo(messageInfo)
                       ?? throw new ArgumentException($"Message type for attribute {messageInfo} is not registered.");
            _subscriptions.Add(type);
            return Task.CompletedTask;
        }

        public bool SupportsCommandMessages => true;

        public bool SupportsDelayedMessages => true;

        public bool SupportsSubscriptions => true;

        public Task Send(Transaction transaction) => transport.Send(transaction);

        public Task Start() => transport.Start();

        public Task Stop() => transport.Stop();
    }

    public async Task Send(Transaction transaction)
    {
        if (!_active)
        {
            throw new InvalidOperationException("Transport is not active.");
        }

        if (transaction.OutgoingMessages.Count == 0)
        {
            return;
        }

        if (Interlocked.Increment(ref _count) == 1)
        {
            while (_emptySignal.CurrentCount > 0)
            {
                await _emptySignal.WaitAsync(0);
            }
        }

        try
        {
            var tasks = new List<Task>();
            var now = DateTime.UtcNow;

            foreach (var message in transaction.OutgoingMessages)
            {
                if (message.DeliverAt != null)
                {
                    var wait = message.DeliverAt.Value - now;
                    if (wait > TimeSpan.Zero)
                    {
                        DelayedSendAsync(message, wait);
                        continue;
                    }
                }

                foreach (var endpoint in _endpoints.Values)
                {
                    var task = endpoint.Handle(message);
                    tasks.Add(task);
                }
            }

            await Task.WhenAll(tasks);
        }
        finally
        {
            if (Interlocked.Decrement(ref _count) == 0)
            {
                _emptySignal.Release();
            }
        }
    }
    async void DelayedSendAsync(OutgoingMessage message, TimeSpan delay)
    {
        try
        {
            await Task.Delay(delay, _cts.Token);
            var transaction = (OutgoingTransaction)await message.Bus.CreateTransaction();
            message.DeliverAt = null;
            transaction.OutgoingMessages.Add(message);
            await Send(transaction);
        }
        catch (OperationCanceledException)
        {
            // Ignore cancellation
        }
        catch (Exception error)
        {
            var logger = message.State.Values.OfType<ILogger>().FirstOrDefault();
            logger?.LogError(error, error.Message);
        }
    }

    public bool UseSubscriptions { get; set; }

    public Task Start()
    {
        _active = true;
        return Task.CompletedTask;
    }
    bool _active;

    public async Task Stop()
    {
        _active = false;
        _cts.Cancel();
        if (Volatile.Read(ref _count) > 0)
        {
            await _emptySignal.WaitAsync();
        }
    }
    int _count;
    readonly CancellationTokenSource _cts = new();
    readonly SemaphoreSlim _emptySignal = new(0, 1);
}
